File: www/gmail.js
var request = require("./request");
var oauth = require("./oauth");

var SEND_ENDPOINT =
  "https://www.googleapis.com/upload/gmail/v1/users/me/messages/send";
var MAX_LINE_B64 = 76;

function token(session, callback, code, refresh) {
  if (typeof code !== "string" || !code) {
    return callback ? callback() : undefined;
  }

  oauth.token(session, oauth.SERVICE.GMAIL, code, refresh, callback);
}

function validateEmail(email) {
  if (typeof email !== "string" || email.length > 100) {
    return;
  }

  // Printable ascii (excluding space)
  if (!new RegExp("^[!-~]{6,100}$").test(email)) {
    return;
  }

  return new RegExp(
      "^[^@=<>;+]+@[a-zA-Z0-9-]{1,64}(\\.[a-zA-Z0-9-]{1,64})*\\.[a-zA-Z]{2,}$")
    .test(email);
}

function send(session, callback, access_token, to, subject, content) {
  if (typeof access_token !== "string" || typeof to !== "string" || typeof subject !==
    "string" ||
    typeof content !== "string" || !validateEmail(to)) {
    return callback ? callback() : undefined;
  }

  var lines = [];

  lines.push("To: <" + to + ">");
  lines.push("Subject: =?UTF-8?B?" + new Buffer(subject).toString("base64") +
    "?=");
  lines.push("Content-Type: text/html; charset=UTF-8");
  lines.push("Content-Transfer-Encoding: base64");
  lines.push("");

  var content_b64 = "";
  var b = new Buffer(content).toString("base64");

  for (var i = 0, c = 0; i < b.length; i++, c++) {
    if (c >= MAX_LINE_B64) {
      content_b64 += "\r\n";
      c = 0;
    }

    content_b64 += b[i];
  }

  lines.push(content_b64);

  var headers = {
    "Authorization": "Bearer " + access_token,
    "Content-Type": "message/rfc822"
  };

  var u = url.parse(SEND_ENDPOINT);

  request("POST", u.hostname, u.pathname + "?uploadType=media", headers, null,
    new Buffer(lines
      .join("\r\n"), "utf8"), callback);
}

module.exports = {
  token: token,
  send: send
};

File: android/src/app/src/main/java/com/example/agitodo/Gmail.java
package com.example.agitodo;

import java.util.ArrayList;
import java.util.List;

import android.net.Uri;
import android.util.Base64;

import java.lang.StringBuffer;

import org.apache.http.message.BasicNameValuePair;
import org.apache.http.NameValuePair;

public class Gmail {
    static final int MAX_EMAIL_LENGTH = 254;
    static final int MIN_EMAIL_LENGTH = 3;
    static final String SEND_ENDPOINT = "https://www.googleapis.com/upload/gmail/v1/users/me/messages/send";

    static Logger log = new Logger();

    String access_token;

    public Gmail(String access_token) {
        this.access_token = access_token;
    }

    public String rawMessage(String to, String subject, String body) {
        StringBuffer lines = new StringBuffer();

        try {
            lines.append("To: <" + to + ">\r\n");
            lines.append("Subject: =?UTF-8?B?"
                    + Base64.encodeToString(subject.getBytes("UTF-8"), Base64.NO_WRAP)
                    + "?=\r\n");
            lines.append("Content-Type: text/html; charset=UTF-8\r\n");
            lines.append("Content-Transfer-Encoding: base64\r\n");
            lines.append("\r\n");
            lines
                    .append(Base64.encodeToString(body.getBytes("UTF-8"), Base64.DEFAULT));
        } catch (Exception e) {
            log.error(e.toString());
            return "";
        }

        return lines.toString();
    }

    public Request.Response send(String to, String subject, String content) {
        List<NameValuePair> headers = new ArrayList<NameValuePair>();
        headers.add(new BasicNameValuePair("Authorization", "Bearer "
                + access_token));
        headers.add(new BasicNameValuePair("Content-Type", "message/rfc822"));

        String body = rawMessage(to, subject, content);

        Uri.Builder uriBuilder = Uri.parse(SEND_ENDPOINT).buildUpon();
        uriBuilder.appendQueryParameter("uploadType", "media");

        return new Request(Request.POST, uriBuilder.build(), headers).send(body);
    }
}
